### DISCUSSION TITLE
Integration: Entry Element: Field Mapping: Is it possible to "add" category ID's to an existing field with Categories?

### DISCUSSION THREAD

**Discussion by adrianjean** [verbb/formie#1943] [Q&A]
I have an entry type called "Member".
On the Member, I have a field for "Awards". It's a category field that allows members to be "tagged" with multiple awards.

I have a Formie form with the Entry Element integration enabled and fields mapped to the fields on my "Member" entry type.

However, on the Formie form I have a dropdown list of awards that people can apply for.

What I'd like to do is allow people to submit the form, and if they selected to add an award to their Member profile, Formie would add it. 

However right now, even though I have selected "Update Entry" as the method for Field Mapping, I notice that Formie overwrites the "awards" field, replacing all the awards on the field with the one award from the Formie form.

Is there a way to tell Formie to add the category, and leave any existing categories in place on the target elements?

**Comment by engram-design** [comment]
I'm afraid not, and this more or less comes down to how Craft handles field values for element fields. It expects you to define _all_ the elements to be included in the elements field. We'd have to introduce settings for element fields in the mapping process to configure this, but we don't have a UI planned for that, despite some [other](https://github.com/verbb/formie/issues/1691) changes planned.

You'd have to either populate the the Formie field you're mapping to your Categories field with all the categories you want, or use an [event](https://verbb.io/craft-plugins/formie/docs/developers/events#the-beforeSendPayload-event) to hook into the save event of the Category via the integration, and grab any existing categories on that field (if an existing category was found) and include their IDs in with the new data coming in Formie.

**Reply by adrianjean** [reply]
Hmmm... I figured as much. 

Could I potentially hook into the `beforeSaveSubmission` event using a custom plugin, and query CraftCMS for my record based on say the email supplied in the Formie form, then see if there are any Categories on that record, add their ID's to the array in the Formie form data?

Or you are suggesting it should be the `beforeSendPayload` event to do all this.

**Reply by engram-design** [reply]
You should use the `beforeSendPayload` for this, as that's going to be the event raised when an element integration runs, and is about to create that Entry element. You'd use this to modify the payload (the field values) for your "awards" field, to include the extra IDs that you want to retain, merging any already mapped (the new ones that a user has picked).

But then again, you could use `beforeSaveSubmission` which is earlier in the process. The benefit of that is that you're updating the values of the submission itself, so when you go to view it in the control panel, you'll see all the categories you've decided to include.
